docs/data/joy/guides/next-js-app-router/next-js-app-router.md

```diff
@@ -0,0 +1,153 @@
+import theme from '/path/to/custom/theme'; // OPTIONAL
+
+// This implementation is from emotion-js
+// https://github.com/emotion-js/emotion/issues/2928#issuecomment-1319747902
+export default function ThemeRegistry(props) {
+  const { options, children } = props;
+
+  const [{ cache, flush }] = React.useState(() => {
+    const cache = createCache(options);
+    cache.compat = true;
+    const prevInsert = cache.insert;
+    let inserted: string[] = [];
+    cache.insert = (...args) => {
+      const serialized = args[1];
+      if (cache.inserted[serialized.name] === undefined) {
+        inserted.push(serialized.name);
+      }
+      return prevInsert(...args);
+    };
+    const flush = () => {
+      const prevInserted = inserted;
+      inserted = [];
+      return prevInserted;
+    };
+    return { cache, flush };
+  });
+
+  useServerInsertedHTML(() => {
+    const names = flush();
+    if (names.length === 0) {
+      return null;
+    }
+    let styles = '';
+    for (const name of names) {
+      styles += cache.inserted[name];
+    }
+    return (
+      <style
+        key={cache.key}
+        data-emotion={`${cache.key} ${names.join(' ')}`}
+        dangerouslySetInnerHTML={{
+          __html: styles,
+        }}
+      />
+    );
+  });
+
+  return (
+    <CacheProvider value={cache}>
+      <CssVarsProvider theme={theme}>
+        {/* the custom theme is optional */}
+        <CssBaseline />
+        {children}
+      </CssVarsProvider>
+    </CacheProvider>
+  );
+}
+
+// app/layout.tsx
+export default function RootLayout({ children }) {
+  return (
+    <html lang="en">
+      <body>
+        <ThemeRegistry options={{ key: 'joy' }}>{children}</ThemeRegistry>
+      </body>
+    </html>
+  );
+}
+```
+
+## Function props
+
+Props passed from server components - for example `page.js` or other routing files - must be [serializable](https://nextjs.org/docs/getting-started/react-essentials#passing-props-from-server-to-client-components-serialization).
+
+This works without any additional directives:
+
+```jsx
+// app/page.tsx
+import Sheet from '@mui/joy/Sheet';
+import Typography from '@mui/joy/Typography';
+
+export default function Page() {
+  return (
+    <>
+      <Sheet variant="outlined">
+        <Typography fontSize="sm">Hello World</Typography>
+      </Sheet>
+    </>
+  );
+}
+```
+
+However, function props such as event handlers or render props are **non-serializable** - this won't work in a server component:
+
+```tsx
+// page.tsx
+import Button from '@mui/joy/Button';
+import Sheet from '@mui/joy/Sheet';
+
+export default function Page() {
+  return (
+    <>
+      <Sheet variant="outlined">
+        {/* Next.js won't render this button without 'use-client' */}
+        <Button
+          variant="outlined"
+          onClick={() => {
+            console.log('handle click');
+          }}
+        >
+          Submit
+        </Button>
+      </Sheet>
+    </>
+  );
+}
+```
+
```

Comment:
I think we should wrap this whole thing in a big `:::error` callout so it's immediately clear that **this does not work**. So it'd be like:

```suggestion
:::error
This code snippet _doesn't work_, because the Button's click handler is non-serializable:

// code 
:::